File: packages/sdk-node/test/webhook.test.ts
import { describe, expect, it } from 'vitest';
import { sign } from '@polaris-mail/hmac';
import { verifyWebhook } from '../src/webhook.js';
import vectors from '../../test-vectors/vectors.json' with { type: 'json' };

interface Vector {
  name: string;
  direction: 'polaris-api' | 'polaris-webhook';
  method: string;
  path: string;
  query: string;
  ts: string;
  nonce: string;
  secret: string;
  body: string;
  expected_sig: string;
  must_verify: boolean;
  expected_error?: string;
}

describe('@polaris/sdk/webhook verifier against shared vectors', () => {
  for (const v of vectors.vectors as Vector[]) {
    it(v.name, async () => {
      const r = await verifyWebhook({
        direction: v.direction,
        method: v.method,
        path: v.path,
        query: v.query,
        headers: {
          'x-polaris-ts': v.ts,
          'x-polaris-nonce': v.nonce,
          'x-polaris-sig': v.expected_sig,
        },
        body: v.body,
        secret: v.secret,
        now: () => Number(v.ts),
      });
      if (v.must_verify) {
        expect(r.ok).toBe(true);
      } else {
        expect(r.ok).toBe(false);
      }
    });
  }
});

// 8g — cross-SDK interop. The same vector is consumed by sdk-go's
// TestCrossSDKInteropPipeQuery (packages/sdk-go/webhook_test.go). Both
// SDKs must produce a byte-identical signature when given a query string
// with percent-encoded `|` characters and a non-empty JSON body — the
// exact divergence-prone shape that has bitten verifiers in other
// projects. Verifying here AND in sdk-go locks the canonical form down
// from both sides.
describe('cross-SDK interop vector', () => {
  it('sdk-node sign() matches the canonical expected_sig', async () => {
    const v = (vectors.vectors as Vector[]).find(
      (x) => x.name === 'interop/POST/query-pipe-encoded-with-body',
    );
    expect(v, 'interop vector missing — regenerate vectors.json').toBeTruthy();
    const got = await sign(
      {
        direction: v!.direction,
        method: v!.method,
        path: v!.path,
        query: v!.query,
        ts: v!.ts,
        nonce: v!.nonce,
        body: v!.body,
      },
      v!.secret,
    );
    expect(got).toBe(v!.expected_sig);
  });
});

describe('un-versioned signature header', () => {
  it('rejects a `v1=…`-prefixed signature outright', async () => {
    const v = (vectors.vectors as Vector[]).find((x) => x.must_verify)!;
    const prefixed = 'v1=' + v.expected_sig;
    const r = await verifyWebhook({
      direction: v.direction,
      method: v.method,
      path: v.path,
      query: v.query,
      headers: {
        'x-polaris-ts': v.ts,
        'x-polaris-nonce': v.nonce,
        'x-polaris-sig': prefixed,
      },
      body: v.body,
      secret: v.secret,
      now: () => Number(v.ts),
    });
    expect(r.ok).toBe(false);
    if (!r.ok) expect(r.code).toBe('invalid_signature');
  });

  it('rejects a `v2=…`-prefixed signature outright', async () => {
    const v = (vectors.vectors as Vector[]).find((x) => x.must_verify)!;
    const prefixed = 'v2=' + v.expected_sig;
    const r = await verifyWebhook({
      direction: v.direction,
      method: v.method,
      path: v.path,
      query: v.query,
      headers: {
        'x-polaris-ts': v.ts,
        'x-polaris-nonce': v.nonce,
        'x-polaris-sig': prefixed,
      },
      body: v.body,
      secret: v.secret,
      now: () => Number(v.ts),
    });
    expect(r.ok).toBe(false);
    if (!r.ok) expect(r.code).toBe('invalid_signature');
  });
});

File: packages/sdk-go/webhook_test.go
package polarissdk

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"testing"
	"time"
)

type vector struct {
	Name          string `json:"name"`
	Direction     string `json:"direction"`
	Method        string `json:"method"`
	Path          string `json:"path"`
	Query         string `json:"query"`
	TS            string `json:"ts"`
	Nonce         string `json:"nonce"`
	Secret        string `json:"secret"`
	Body          string `json:"body"`
	ExpectedSig   string `json:"expected_sig"`
	MustVerify    bool   `json:"must_verify"`
	ExpectedError string `json:"expected_error"`
}

type vectorsFile struct {
	Vectors []vector `json:"vectors"`
}

func loadVectors(t *testing.T) []vector {
	t.Helper()
	_, here, _, _ := runtime.Caller(0)
	root := filepath.Join(filepath.Dir(here), "..", "test-vectors", "vectors.json")
	data, err := os.ReadFile(root)
	if err != nil {
		t.Fatalf("read vectors: %v", err)
	}
	var f vectorsFile
	if err := json.Unmarshal(data, &f); err != nil {
		t.Fatalf("parse vectors: %v", err)
	}
	return f.Vectors
}

func TestVerifyAgainstSharedVectors(t *testing.T) {
	for _, v := range loadVectors(t) {
		v := v
		t.Run(v.Name, func(t *testing.T) {
			tsMs, _ := strconv.ParseInt(v.TS, 10, 64)
			res := VerifyWebhook(VerifyInput{
				Direction: Direction(v.Direction),
				Method:    v.Method,
				Path:      v.Path,
				Query:     v.Query,
				Headers: map[string]string{
					"x-polaris-ts":    v.TS,
					"x-polaris-nonce": v.Nonce,
					"x-polaris-sig":   v.ExpectedSig,
				},
				Body:   []byte(v.Body),
				Secret: []byte(v.Secret),
				Now:    time.UnixMilli(tsMs),
			})
			if v.MustVerify && !res.OK {
				t.Fatalf("expected ok: %+v", res)
			}
			if !v.MustVerify && res.OK {
				t.Fatalf("expected fail: %+v", res)
			}
		})
	}
}

// TestCrossSDKInteropPipeQuery proves sdk-go produces a signature that is
// byte-identical to the canonical expected_sig for the
// `interop/POST/query-pipe-encoded-with-body` vector. The same vector is
// consumed by sdk-node's vector-driven verifier (packages/sdk-node/test/
// webhook.test.ts), so a divergence in canonicalization between the two
// SDKs would surface here OR there. Together they form an end-to-end
// interop guarantee: any signer in either language can be verified by the
// other. (8g)
func TestCrossSDKInteropPipeQuery(t *testing.T) {
	var v vector
	for _, x := range loadVectors(t) {
		if x.Name == "interop/POST/query-pipe-encoded-with-body" {
			v = x
			break
		}
	}
	if v.Name == "" {
		t.Fatal("interop vector missing — regenerate vectors.json")
	}
	got, err := Sign(CanonicalInput{
		Direction: Direction(v.Direction),
		Method:    v.Method,
		Path:      v.Path,
		Query:     v.Query,
		TS:        v.TS,
		Nonce:     v.Nonce,
		Body:      []byte(v.Body),
	}, []byte(v.Secret))
	if err != nil {
		t.Fatalf("Sign: %v", err)
	}
	if got != v.ExpectedSig {
		t.Fatalf("sdk-go Sign() diverged from canonical:\n  got  = %s\n  want = %s", got, v.ExpectedSig)
	}
}

func TestRejectsVersionedSignaturePrefix(t *testing.T) {
	vs := loadVectors(t)
	var v vector
	for _, x := range vs {
		if x.MustVerify {
			v = x
			break
		}
	}
	// Prefix the otherwise-valid signature with `v1=` — must be rejected.
	prefixed := "v1=" + v.ExpectedSig
	tsMs, _ := strconv.ParseInt(v.TS, 10, 64)
	res := VerifyWebhook(VerifyInput{
		Direction: Direction(v.Direction),
		Method:    v.Method,
		Path:      v.Path,
		Query:     v.Query,
		Headers: map[string]string{
			"x-polaris-ts":    v.TS,
			"x-polaris-nonce": v.Nonce,
			"x-polaris-sig":   prefixed,
		},
		Body:   []byte(v.Body),
		Secret: []byte(v.Secret),
		Now:    time.UnixMilli(tsMs),
	})
	if res.OK {
		t.Fatalf("expected fail for prefixed sig, got: %+v", res)
	}
	if res.Code != CodeInvalidSignature {
		t.Fatalf("expected code=invalid_signature, got %s", res.Code)
	}
}
